New plugin: computed fields.

jon-heard opened the pull request.

jon-heard committed:
New plugin: computed fields.

seleb commented on src/computed fields.js:
```diff
@@ -0,0 +1,161 @@
+		field for the graphic, while the playback will use the field of step 4 for the graphic.
+
+
+HOW TO USE - COMPUTE TERRAIN PASSABILITY BASED ON WHETHER AN ITEM IS TAKEN
+1. Import this plugin into a game with an open room to move about in.
+2. Draw a "boots" tile for an item.
+3. Draw a "spikes" tile for the floor.
+4. Add a "message" event beside the player.  Set the "say" field to "I put on the boots.".  Add a tag
+		field and name it "boots".  Add a tile field named "graphic" and set it to the boots tile of step 2.
+5. Add a "character" event a few cells from the avatar.  Set its graphic to the spikes tile of step 3.  Erase the "say" field.
+6. Change the "solid" tag field for the event of step 5 to javascript type.  Set it to this code:
+  if (!FIND_EVENT("boots")) return;
+  return { type: "tag", data: true };
+7. Copy the event of step 5.  Paste it around the player and the boots so that they are surrounded.
+8. Playtest the game.  Try to step over the spikes without getting the boots.  You should be blocked.  Walk over the boots and now the spikes should not block you.
```
i would strongly recommend rewriting this example in terms of a locked door + key, which is what people tend to ask how to do

seleb commented on src/computed fields.js:
```diff
@@ -0,0 +1,161 @@
+8. Playtest the game.  Try to step over the spikes without getting the boots.  You should be blocked.  Walk over the boots and now the spikes should not block you.
+
+
+HOW TO USE - DIALOGUE CHANGES BASED ON HOW LONG THE GAME IS PLAYED
+1. Import this plugin into a game.
+2. Add a character event near the avatar with a custom graphic tile.
+3. Set the "say" field for the event of step 2 to a javascript type.  Set it to this code:
+  CLEAR_EVENT_SAY_CACHE(EVENT);
+  let data =
+    "We've been here for " +
+    PLAYBACK.frameCount +
+    " ticks.";
+  return { type: "dialogue", data: data };
+4. Run the game and repeatedly interact with the event of step 2.  Note that the dialogue changes.
+5. Also, not that the code begins with "CLEAR_EVENT_SAY_CACHE(EVENT);".  This is important to render
+	the latest "say" value.  Without it the dialogue is cached, so won't show changes.
```
this isn't really a practical example of this plugin: the benefit of this plugin is being able to write declarative logic for state changes that would ordinarily require imperatively modifying events, but bipsi already lets you script dialogue in this way from the regular js fields by calling `SAY`.

i'd recommend removing all the `say`-specific parts of this plugin since it's redundant with the engine features

seleb commented on src/computed fields.js:
```diff
@@ -0,0 +1,161 @@
+};
+
+// A variation of "runJS" which (A) is synchronous (B) returns the return-value of the code.
+BipsiPlayback.prototype.runJSNonAsync = function runJSNonAsync(event, js, debug = false) {
```
"non async" is a double-negative; you should call this `runJSSync`

jon-heard committed:
Renamed method - "runJSNonAsync" to "runJSSync"
Improved HOW-TO sections
Removed CLEAR_EVENT_SAY_CACHE as the technique it was for is not relevant.

jon-heard committed:
Fixed comment indenting

jon-heard commented:
I've made the requested updates.  The key/door HOW-TO is longer than the boot/spikes HOW-TO, but it's more fun now.

seleb committed:
Merge branch 'main' into newPlugin_calculatedFields

seleb commented:
:tada: This PR is included in version 4.10.0 :tada:

The release is available on [GitHub release](https://github.com/seleb/bipsi-hacks/releases/tag/v4.10.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
